app/service/auth_svc.py

```diff
@@ -88,21 +96,40 @@ async def login_user(self, request):
         :param request:
         :return: the response/location of where the user is trying to navigate
         """
-        data = await request.post()
-        username = data.get('username')
-        password = data.get('password')
-        if self.ldap_config:
-            verified = await self._ldap_login(username, password)
-        else:
-            verified = await self._check_credentials(request.app.user_map, username, password)
-
-        if verified:
-            self.log.debug('%s logging in:' % username)
-            response = web.HTTPFound('/')
-            await remember(request, response, username)
-            raise response
-        self.log.debug('%s failed login attempt: ' % username)
-        raise web.HTTPFound('/login')
+        try:
+            self.log.debug('Using login handler "%s" for login' % self._login_handler.name)
+            await self._login_handler.handle_login(request)
+        except web.HTTPRedirection as http_redirect:
+            raise http_redirect
+        except web.HTTPUnauthorized as http_unauthorized:
+            raise http_unauthorized
+        except Exception as e:
+            self.log.error('Exception when handling login request: %s' % e)
+
+        # Fallback if not already using default login handler
+        if not isinstance(self._login_handler, self.DefaultLoginHandler):
+            self.log.debug('Falling back to default login handler')
+            return await self._default_login_handler.handle_login(request)
+
+    async def login_redirect(self, request, use_template=True):
+        """Redirect user to login page using the configured login handler. If using the default login handler
+        and use_template is set to true, method will return the login.html template. If use_template is set to False
+        and the default login handler is configured, it will redirect to '/login' by raising HTTPFound exception."""
+
+        try:
+            self.log.debug('Using login handler "%s" for login redirect' % self._login_handler.name)
+            return await self._login_handler.handle_login_redirect(request, use_template=use_template)
+        except web.HTTPRedirection as http_redirect:
+            raise http_redirect
+        except web.HTTPUnauthorized as http_unauthorized:
+            raise http_unauthorized
```

Comment:
Does this cover all the exception response types we care about? https://docs.aiohttp.org/en/stable/web_quickstart.html#aiohttp-web-exceptions

Reply:
It might make sense to include HTTPSuccessful and HTTPForbidden as potential expected exceptions that we want to raise to the caller.

Reply:
I went ahead and expanded the number of exceptions that we "allow"